src/modules/Client/Api/Guest.php

```diff
@@ -150,78 +151,99 @@ public function login($data)
         if (!$c instanceof \Model_Client) {
             return true;
         }
 
-        $hash = hash('sha256', time() . random_bytes(13));
+        // Check if a password reset request exists
+        $reset = $this->di['db']->findOne('ClientPasswordReset', 'client_id = ?', [$c->id]);
 
-        $reset = $this->di['db']->dispense('ClientPasswordReset');
-        $reset->client_id = $c->id;
-        $reset->ip = $this->ip;
-        $reset->hash = $hash;
-        $reset->created_at = date('Y-m-d H:i:s');
-        $reset->updated_at = date('Y-m-d H:i:s');
-        $this->di['db']->store($reset);
+        // If no recent reset request exists, create a new one
+        if (!$reset instanceof \Model_ClientPasswordReset) {
+            $hash = hash('sha256', time() . random_bytes(13));
+            $reset = $this->di['db']->dispense('ClientPasswordReset');
+            $reset->client_id = $c->id;
+            $reset->ip = $this->ip;
+            $reset->hash = $hash;
+            $reset->created_at = date('Y-m-d H:i:s');
+            $reset->updated_at = date('Y-m-d H:i:s');
+            $this->di['db']->store($reset);
+        }
 
-        // send email
-        $email = [];
-        $email['to_client'] = $c->id;
-        $email['code'] = 'mod_client_password_reset_request';
-        $email['hash'] = $hash;
+        // prepare reset email
+        $email = [
+            'to_client' => $c->id,
+            'code' => 'mod_client_password_reset_request',
+            'hash' => $reset->hash,
+        ];
+        // Send the email if the reset request has the same created_at and updated_at
         $emailService = $this->di['mod_service']('email');
-        $emailService->sendTemplate($email);
+        if ($reset->created_at == $reset->updated_at) {
+            $emailService->sendTemplate($email);
+        } elseif (strtotime($reset->updated_at) - time() + 60 < 0) {
+            $emailService->sendTemplate($email);
+        }
+        // update the client password reset time 
+        $reset->updated_at = date('Y-m-d H:i:s');
+        $this->di['db']->store($reset);
 
         $this->di['logger']->info('Client requested password reset. Sent to email %s', $c->email);
 
         return true;
     }
 
-    /**
-     * Confirm password reset action.
-     *
-     * @return bool
-     *
-     * @throws \Box_Exception
-     */
-    public function confirm_reset($data)
+
+    public function update_password($data)
     {
         $required = [
-            'hash' => 'Hash required',
+            'hash' => 'No Hash provided',
+            'password' => 'Password required',
+            'password_confirm' => 'Password confirmation required',
         ];
-        $this->di['events_manager']->fire(['event' => 'onBeforePasswordResetClient']);
-        $this->di['validator']->checkRequiredParamsForArray($required, $data);
+        $this->di['events_manager']->fire(['event' => 'onBeforeClientProfilePasswordSet', 'params' => $data['hash']]);
```

Comment:
This should really be `onBeforeClientProfilePasswordChange` (an already existing event that's in use & the "after" variant is already being sent at this end of this function)

Reply:
I had that before, but then I though - maybe we _don't_ want to use the same event, because ```onBeforeClientProfilePasswordChange``` is used when the user manually changes their password from a logged in state, and ```onBeforeClientProfilePasswordSet``` happens when the user resets their password. Maybe we want to differentiate between these two events for auditing purposes?

Reply:
I'm indifferent to the two, but I think `onBeforeClientProfilePasswordReset` makes more sense if we are to implement a new one.
And if we are going to implement one, we should fire an "after" variant of it as well (It's not technically required, just generally how everything else is done and I'd like to keep it consistent)

Reply:
I've renamed it.